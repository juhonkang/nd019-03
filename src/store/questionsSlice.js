import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { _getQuestions, _saveQuestion, _saveQuestionAnswer } from '../../_DATA.js';

export const fetchQuestions = createAsyncThunk('questions/fetchQuestions', async () => {
  const questions = await _getQuestions();
  return questions;
});

export const saveQuestion = createAsyncThunk('questions/saveQuestion', async (question) => {
  const savedQuestion = await _saveQuestion(question);
  return savedQuestion;
});

export const saveQuestionAnswer = createAsyncThunk('questions/saveQuestionAnswer', async (answerData) => {
  await _saveQuestionAnswer(answerData);
  return answerData;
});

const questionsSlice = createSlice({
  name: 'questions',
  initialState: {
    questions: {},
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchQuestions.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchQuestions.fulfilled, (state, action) => {
        state.loading = false;
        state.questions = action.payload;
      })
      .addCase(fetchQuestions.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      })
      .addCase(saveQuestion.fulfilled, (state, action) => {
        state.questions[action.payload.id] = action.payload;
      })
      .addCase(saveQuestionAnswer.fulfilled, (state, action) => {
        const { qid, answer, authedUser } = action.payload;
        if (state.questions[qid]) {
          state.questions[qid][answer].votes.push(authedUser);
        }
      });
  },
});

export default questionsSlice.reducer;