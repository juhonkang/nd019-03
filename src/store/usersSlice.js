import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { _getUsers } from '../../_DATA.js';

export const fetchUsers = createAsyncThunk('users/fetchUsers', async () => {
  const users = await _getUsers();
  return users;
});

const usersSlice = createSlice({
  name: 'users',
  initialState: {
    users: {},
    loading: false,
    error: null,
  },
  reducers: {
    updateUserAnswer: (state, action) => {
      const { userId, questionId, answer } = action.payload;
      if (state.users[userId]) {
        state.users[userId].answers[questionId] = answer;
      }
    },
    addUserQuestion: (state, action) => {
      const { userId, questionId } = action.payload;
      if (state.users[userId]) {
        state.users[userId].questions.push(questionId);
      }
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchUsers.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchUsers.fulfilled, (state, action) => {
        state.loading = false;
        state.users = action.payload;
      })
      .addCase(fetchUsers.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      });
  },
});

export const { updateUserAnswer, addUserQuestion } = usersSlice.actions;
export default usersSlice.reducer;