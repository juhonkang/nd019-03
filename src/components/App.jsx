import React, { useEffect } from 'react';
import { BrowserRouter as Router, Routes, Route, Navigate } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { fetchUsers } from '../store/usersSlice';
import { fetchQuestions } from '../store/questionsSlice';
import Navigation from './Navigation.jsx';
import Login from './Login.jsx';
import Home from './Home.jsx';
import PollDetail from './PollDetail.jsx';
import NewPoll from './NewPoll.jsx';
import Leaderboard from './Leaderboard.jsx';
import NotFound from './NotFound.jsx';
import ProtectedRoute from './ProtectedRoute.jsx';

function App() {
  const dispatch = useDispatch();
  const { isAuthenticated } = useSelector((state) => state.auth);

  useEffect(() => {
    dispatch(fetchUsers());
    dispatch(fetchQuestions());
  }, [dispatch]);

  return (
    <Router>
      <div className="app">
        {isAuthenticated && <Navigation />}
        <main className="main-content">
          <Routes>
            <Route path="/login" element={<Login />} />
            <Route path="/" element={
              <ProtectedRoute>
                <Home />
              </ProtectedRoute>
            } />
            <Route path="/questions/:id" element={
              <ProtectedRoute>
                <PollDetail />
              </ProtectedRoute>
            } />
            <Route path="/add" element={
              <ProtectedRoute>
                <NewPoll />
              </ProtectedRoute>
            } />
            <Route path="/leaderboard" element={
              <ProtectedRoute>
                <Leaderboard />
              </ProtectedRoute>
            } />
            <Route path="/404" element={<NotFound />} />
            <Route path="*" element={<Navigate to="/404" replace />} />
          </Routes>
        </main>
      </div>
    </Router>
  );
}

export default App;
