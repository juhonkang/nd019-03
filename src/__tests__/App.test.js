import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import App from '../components/App.jsx';
import authSlice from '../store/authSlice';
import usersSlice from '../store/usersSlice';
import questionsSlice from '../store/questionsSlice';

const createMockStore = () => {
  return configureStore({
    reducer: {
      auth: authSlice,
      users: usersSlice,
      questions: questionsSlice,
    },
    preloadedState: {
      auth: { user: null, isAuthenticated: false },
      users: { users: {}, loading: false, error: null },
      questions: { questions: {}, loading: false, error: null }
    }
  });
};

describe('App Component', () => {
  it('should render without crashing', () => {
    const store = createMockStore();
    
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
  });

  it('should match snapshot', () => {
    const store = createMockStore();
    
    const { container } = render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    
    expect(container.firstChild).toMatchSnapshot();
  });
});