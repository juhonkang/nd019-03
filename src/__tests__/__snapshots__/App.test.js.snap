// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match snapshot 1`] = `
<div
  class="app"
>
  <main
    class="main-content"
  >
    <div
      class="loading"
    >
      Loading...
    </div>
  </main>
</div>
`;
